exercises/enumerating_trees/descr.md

```diff
@@ -0,0 +1,364 @@
+Its implementation is the topic of other exercises, entitled
+"Symbolic Sequences as Objects" and
+"Symbolic Sequences as Data".
+
+## The `Seq` API
+
+`'a seq` is the type of a sequence whose elements have type `'a`.
+```
+  type 'a seq
+```
+
+### Constructors
+
+There are five constructor functions:
+`empty`,
+`singleton`,
+`sum`,
+`product`,
+and `map`.
+
+Each of these functions has **constant time** complexity.
+Indeed, they do not actually construct the entire sequence in memory;
+they construct only a compact representation of it.
+
+The constant `empty` is the empty sequence.
+
+```
+  val empty: 'a seq
+```
+
+The sequence `singleton x` has length 1. Its single element is `x`.
+
+```
+  val singleton: 'a -> 'a seq
+```
+
+The sequence `sum s1 s2` is the concatenation of the sequences `s1` and `s2`.
+
+```
+  val sum: 'a seq -> 'a seq -> 'a seq
+```
+
+The sequence `bigsum ss` is the concatenation of all sequences in the list
+`ss`.
+
+```
+  val bigsum: 'a seq list -> 'a seq
+```
+
+The sequence `product s1 s2`, a sequence of pairs, is the Cartesian product
+of the sequences `s1` and `s2`. Its length is the product of the lengths of
+`s1` and `s2`. The first pair component is considered most significant,
+which means that the product of the integer sequence `0; 1` by itself is the
+sequence `(0, 0); (0, 1); (1, 0); (1, 1)`.
+
+```
+  val product: 'a seq -> 'b seq -> ('a * 'b) seq
+```
+
+The sequence `map phi s` is the image of the sequence `s` through the function
+`phi`.
+
+```
+  val map: ('a -> 'b) -> 'a seq -> 'b seq
+```
+
+### Destructors
+
+There are three destructors: `length`, `get`, and `foreach`.
+
+`length s` returns the length of the sequence `s`.
+
+```
+  val length: 'a seq -> int
+```
+
+If the index `i` is comprised between zero (included) and `length s` (excluded),
+then `get s i` returns the `i`-th element of the sequence `s`.
+Otherwise, `get s i` raises the exception `OutOfBounds`.
+
+```
+  exception OutOfBounds
+  val get: 'a seq -> int -> 'a
+```
+
+`foreach s k` iterates over all elements of the sequence `s`.
+Each element in turn is passed to the loop body `k`.
+Thus, the OCaml code `foreach s (fun x -> ...)` can be
+informally understood as `for each x in s do ...`.
+
+```
+  val foreach: 'a seq -> ('a -> unit) -> unit
+```
+
+`elements s` returns a list of all elements of the sequence `s`.
+In principle, you don't need to use it,
+but it can be useful for debugging.
+
+```
+  val elements: 'a seq -> 'a list
+```
+
+## Constructing enumerations
+
+We begin by defining six basic functions that construct enumerations.
+
+**Question 1.** Define a constant `empty` of type `'a enum` which represents
+the empty set.
+
+**Question 2.** Define a function `just` of type `'a -> 'a enum` such that
+`just x` represents a singleton set, whose element is `x`, and the size of
+this element is considered to be zero.
+
+**Question 3.** Define a function `pay` of type `'a enum -> 'a enum` such that
```

Comment:
Can you give a bit of context to explain the naming choice for this function?

Reply:
Ok. I have added the following:
*Note*. `pay` is so named because the elements of the enumeration `pay e` are
*more expensive*, so to speak, than the elements of the enumeration `e`.

Reply:
Did you push these changes?

Reply:
Nope, as I said, I am having trouble because the repository upon which this PR was based has disappeared.